batch/batch/log_store.py

```diff
@@ -11,50 +11,76 @@
 
 
 class LogStore:
-    def __init__(self, batch_logs_bucket_name, instance_id, blocking_pool, *, project=None, credentials=None):
+    def __init__(self, batch_logs_bucket_name, instance_id, *, credentials=None):
         self.batch_logs_bucket_name = batch_logs_bucket_name
         self.instance_id = instance_id
         self.batch_logs_root = f'gs://{batch_logs_bucket_name}/batch/logs/{instance_id}/batch'
-        self.gcs = GCS(blocking_pool, project=project, credentials=credentials)
+
+        self.gcs_fs = GoogleStorageAsyncFS(credentials=credentials)
+        self.sem = asyncio.Semaphore(50)
 
         log.info(f'BATCH_LOGS_ROOT {self.batch_logs_root}')
         format_version = BatchFormatVersion(BATCH_FORMAT_VERSION)
         log.info(f'EXAMPLE BATCH_JOB_LOGS_PATH {self.log_path(format_version, 1, 1, "abc123", "main")}')
 
+    async def _write(self, url, data, binary=True):
+        async with await self.gcs_fs.create(url) as f:
+            if not binary:
+                data = data.encode('utf-8')
+            await f.write(data)
+
+    async def _read(self, url, binary=True):
+        async with await self.gcs_fs.open(url) as f:
+            data = await f.read()
+            if not binary:
+                data = data.decode('utf-8')
+            return data
+
+    async def _rmtree(self, url):
+        async with self.sem:
+            await self.gcs_fs.rmtree(self.sem, url)
```

Comment:
Why is there only a semaphore on rmtree?

Reply:
Because Cotton removes files in parallel. But you have to have a maximum parallelism for that. The other file system operations aren't in parallel.